python/lsst/dbb/buffmngrs/handoff/remote.py

```diff
@@ -118,104 +121,163 @@ def run(self):
-            chunk = []
-            for _ in range(self.chunk_size):
-                try:
-                    topdir, subdir, file = self.todo.get(block=False)
-                except queue.Empty:
-                    break
-                else:
-                    chunk.append((topdir, subdir, file))
-            if not chunk:
+            # Grab a bunch of file items from the input queue.
+            files = get_chunk(self.todo, size=self.chunk_size)
+            if not files:
                 continue
 
             # Group files based on their location as only the files sharing the
             # same location can be transferred as a group with a single
             # transfer command when the batch mode is enabled.
             mapping = {}
-            for topdir, subdir, filename in chunk:
-                mapping.setdefault((topdir, subdir), []).append(filename)
+            for item in files:
+                head, tail, *rest = dataclasses.astuple(item)
+                mapping.setdefault((head, tail), []).append(item)
 
             # Transfer files to the handoff site to the endpoint site.
-            for location, filenames in mapping.items():
+            for location, files in mapping.items():
                 head, tail = location
 
-                # Transfer files to the staging area on the endpoint site.
+                filenames = [item.name for item in files]
+                sizes = {item.name: item.size for item in files}
+
+                # Divide files into batches. If batch mode is enabled,
+                # all files grouped into a single batch. Otherwise,
+                # each batch will consist of a single file.
+                batch_size = len(filenames) if self.batch_mode else 1
+                batches = [filenames[i:i+batch_size]
+                           for i in range(0, len(filenames), batch_size)]
+
+                # Create corresponding number of transfer items to put in the
+                # output queue.
+                transfers = [TransferMsg() for _ in batches]
+                for batch, transfer in zip(batches, transfers):
+                    transfer.files = tuple((head, tail, fn) for fn in batch)
+                    transfer.size = sum(sizes[fn] for fn in batch)
+
+                # 1. PRE-TRANSFER actions
+                # -----------------------
                 dest = os.path.join(stage, tail)
-                relocated = collections.deque()
+                relocated = []
+                total = datetime.timedelta()
 
+                # Create a relevant subdirectory in the staging area.
                 tpl = self.cmds["remote"]
                 cmd = tpl.format(**self.params, command=f"mkdir -p {dest}")
-                status, stdout, stderr = execute(cmd, timeout=self.timeout)
+                start = datetime.datetime.now()
+                status, stdout, stderr, dur = \
+                    execute(cmd, timeout=self.timeout)
+                total += dur
```

Comment:
You're right, it's not really required here. After all, the pre-transfer phase really consists of just one action.

I just implemented it this way to keep things consistent with calculating commands' duration in post-transfer phase (where there are actually two actions) and "made room" for changes in case the pre-transfer phase would ever require more than one action too.

If you think it's an overkill, I will gladly change it.

Reply:
Please change